File: applicatie/src/main/scala/com/vdzon/administratie/checkandfix/actions/check/BestaanCheck.scala
package com.vdzon.administratie.checkandfix.actions.check

import com.vdzon.administratie.checkandfix.CheckAndFixData
import com.vdzon.administratie.checkandfix.model.{CheckType, CheckAndFixRegel, FixAction}
import com.vdzon.administratie.model.boekingen.relaties.{BoekingMetFactuur, BoekingMetRekening, BoekingMetAfschrift}

object BestaanCheck {

  def checkOfAfschriftNogBestaat(data: CheckAndFixData) =
    data.alleBoekingen
      .filter(boeking => boeking match {
        case b:BoekingMetAfschrift => true
        case _ => false
      })
      .map(boeking => boeking.asInstanceOf[BoekingMetAfschrift])
      .filter(boeking => !afschriftExists(boeking.getAfschriftNummer, data))
      .map(boeking => CheckAndFixRegel.newBuilder
        .rubriceerAction(FixAction.REMOVE_BOEKING)
        .checkType(CheckType.FIX_NEEDED)
        .boekingUuid(boeking.getUuid)
        .omschrijving("Afschift " + boeking.getAfschriftNummer + " bestaat niet meer terwijl er wel een boeking van bestaat")
        .build)

  def checkOfRekeningNogBestaat(data: CheckAndFixData) =
    data.alleBoekingen
      .filter(boeking => boeking match {
        case b:BoekingMetRekening => true
        case _ => false
      })
      .map(boeking => boeking.asInstanceOf[BoekingMetRekening])
      .filter(boeking => !rekeningExists(boeking.getRekeningNummer, data))
      .map(boeking => CheckAndFixRegel.newBuilder
        .rubriceerAction(FixAction.REMOVE_BOEKING)
        .checkType(CheckType.FIX_NEEDED)
        .boekingUuid(boeking.getUuid)
        .omschrijving("Rekening " + boeking.getRekeningNummer + " bestaat niet meer terwijl er wel een boeking van bestaat")
        .build)

  def checkOfFactuurNogBestaat(data: CheckAndFixData) =
    data.alleBoekingen
      .filter(boeking => boeking match {
        case b:BoekingMetFactuur => true
        case _ => false
      })
      .map(boeking => boeking.asInstanceOf[BoekingMetFactuur])
      .filter(boeking => !factuurExists(boeking.getFactuurNummer, data))
      .map(boeking => CheckAndFixRegel.newBuilder
        .rubriceerAction(FixAction.REMOVE_BOEKING)
        .checkType(CheckType.FIX_NEEDED)
        .boekingUuid(boeking.getUuid)
        .omschrijving("Factuur " + boeking.getFactuurNummer + " bestaat niet meer terwijl er wel een boeking van bestaat")
        .build)

  private def afschriftExists(afschriftNummer: String, data: CheckAndFixData): Boolean = {
    return data.alleAfschriften.filter(afschrift => afschrift.getNummer().equals(afschriftNummer)).size > 0
  }

  private def rekeningExists(rekeningNummer: String, data: CheckAndFixData): Boolean = {
    return data.alleRekeningen.filter(rekening => rekening.getRekeningNummer().equals(rekeningNummer)).size > 0
  }

  private def factuurExists(factuurNummer: String, data: CheckAndFixData): Boolean = {
    return data.alleFacturen.filter(factuur => factuur.getFactuurNummer().equals(factuurNummer)).size > 0
  }
}

File: checkandfix/impl/src/main/java/com/vdzon/administratie/checkandfix/actions/check/BestaanCheck.kt
package com.vdzon.administratie.checkandfix.actions.check

import com.vdzon.administratie.checkandfix.CheckAndFixData
import com.vdzon.administratie.checkandfix.CheckAndFixRegel
import com.vdzon.administratie.checkandfix.model.CheckType
import com.vdzon.administratie.checkandfix.model.FixAction
import com.vdzon.administratie.model.boekingen.relaties.BoekingMetAfschrift
import com.vdzon.administratie.model.boekingen.relaties.BoekingMetFactuur
import com.vdzon.administratie.model.boekingen.relaties.BoekingMetRekening


object BestaanCheck {
    fun checkOfAfschriftNogBestaat(data: CheckAndFixData): List<CheckAndFixRegel> =
            data.alleBoekingen
                    .filter { boeking ->
                        if (boeking is BoekingMetAfschrift) true else false
                    }
                    .map { boeking -> boeking as BoekingMetAfschrift }
                    .filter { boeking -> !afschriftExists(boeking.afschriftNummer, data) }
                    .map {
                        boeking ->
                        CheckAndFixRegel(
                                rubriceerAction = FixAction.REMOVE_BOEKING,
                                checkType = CheckType.FIX_NEEDED,
                                boekingUuid = boeking.uuid,
                                omschrijving = "Afschift " + boeking.afschriftNummer + " bestaat niet meer terwijl er wel een boeking van bestaat")
                    }


    fun checkOfRekeningNogBestaat(data: CheckAndFixData): List<CheckAndFixRegel> =
            data.alleBoekingen
                    .filter { boeking -> if (boeking is BoekingMetRekening) true else false }
                    .map { boeking -> boeking as BoekingMetRekening }
                    .filter { boeking -> !rekeningExists(boeking.rekeningNummer, data) }
                    .map {
                        boeking ->
                        CheckAndFixRegel(
                                rubriceerAction = FixAction.REMOVE_BOEKING,
                                checkType = CheckType.FIX_NEEDED,
                                boekingUuid = boeking.uuid,
                                omschrijving = "Rekening " + boeking.rekeningNummer + " bestaat niet meer terwijl er wel een boeking van bestaat")
                    }

    fun checkOfFactuurNogBestaat(data: CheckAndFixData): List<CheckAndFixRegel> =
            data.alleBoekingen
                    .filter { boeking -> if (boeking is BoekingMetFactuur) true else false }
                    .map { boeking -> boeking as BoekingMetFactuur }
                    .filter { boeking -> !factuurExists(boeking.factuurNummer, data) }
                    .map { boeking ->
                        CheckAndFixRegel(
                                rubriceerAction = FixAction.REMOVE_BOEKING,
                                checkType = CheckType.FIX_NEEDED,
                                boekingUuid = boeking.uuid,
                                omschrijving = "Factuur " + boeking.factuurNummer + " bestaat niet meer terwijl er wel een boeking van bestaat")
                    }

    private fun afschriftExists(afschriftNummer: String, data: CheckAndFixData): Boolean =
            data.alleAfschriften.filter { afschrift -> afschrift.nummer.equals(afschriftNummer) }.size > 0

    private fun rekeningExists(rekeningNummer: String, data: CheckAndFixData): Boolean =
            data.alleRekeningen.filter { rekening -> rekening.rekeningNummer.equals(rekeningNummer) }.size > 0

    private fun factuurExists(factuurNummer: String, data: CheckAndFixData): Boolean =
            data.alleFacturen.filter { factuur -> factuur.factuurNummer.equals(factuurNummer) }.size > 0


}
